api/saml/metadata/filter.py

```diff
@@ -0,0 +1,131 @@
+        :type inner_exception: Exception
+        """
+        message = "Incorrect filter expression: {0}".format(str(inner_exception))
+
+        super(SAMLSubjectFilterError, self).__init__(message, inner_exception)
+
+
+class SAMLDummyAttributesDict(dict):
+    """Dummy class used for validating SAML filtration expressions.
+
+    This class always returns a SAML attribute with the same name as requested and signle "dummy" value.
+    """
+
+    def __init__(self):
+        """Initialize a new instance of SAMLDummyAttributesDict."""
+        super(SAMLDummyAttributesDict, self).__init__()
+
+    def __getitem__(self, attribute_name):
+        """Return a SAMLAttribute object by its name.
+
+        :param attribute_name: SAML attribute's name
+        :type attribute_name: str
+
+        :return: Dummy SAML attribute name with the same name as requested and single "dummy" value
+        :rtype: SAMLAttribute
+        """
+        return SAMLAttribute(attribute_name, ["dummy"])
+
+
+class SAMLDummyAttributeStatement(SAMLAttributeStatement):
+    """Dummy class used for validating SAML filtration expressions.
+
+    This attribute statement contains a SAMLDummyAttributesDict instance returning dummy SAML attributes.
+    """
+
+    def __init__(self):
+        """Initialize a new instance of SAMLDummyAttributeStatement class."""
+        super(SAMLDummyAttributeStatement, self).__init__([])
+
+        self._attributes = SAMLDummyAttributesDict()
+
+    @property
+    def attributes(self):
+        """Return a SAMLDummyAttributesDict object.
+
+        :return: SAMLDummyAttributesDict object
+        :rtype: SAMLDummyAttributesDict
+        """
+        return self._attributes
+
+
+class SAMLSubjectFilter(object):
+    """Executes filter expressions."""
+
+    def __init__(self):
+        """Initialize a new instance of SAMLSubjectFilter class."""
+        self._logger = logging.getLogger(__name__)
+
+    def execute(self, expression, subject):
+        """Apply the expression to the subject and return a boolean value indicating whether it's a valid subject.
+
+        :param expression: String containing the filter expression
+        :type expression: str
+
+        :param subject: SAML subject
+        :type subject: api.saml.metadata.model.SAMLSubject
+
+        :return: Boolean value indicating whether it's a valid subject
+        :rtype: bool
+
+        :raise SAMLSubjectFilterError: in the case of any errors occurred during expression evaluation
+        """
+        if not expression or not is_string(expression):
+            raise ValueError("Argument 'expression' must be a non-empty string")
+        if not isinstance(subject, SAMLSubject):
+            raise ValueError("Argument 'subject' must an instance of Subject class")
+
+        self._logger.info(
+            "Started applying expression '{0}' to {1}".format(expression, subject)
+        )
+
+        try:
+            result = eval(expression)
```

Comment:
This is what I was afraid of. Is possible to represent these filters as XPath expressions instead of Python code?

Reply:
The problem is that we don't know how complex can these filter expression can be: depending on the university they might want to filter based on several attributes or their combinations using plain mathematical operators or regular expressions.
I see the following alternatives:
- Add additional UI controls to `circulation-web` which would allow to create complex expression using disjunction, conjunction, and composition
- Change `eval` to [`asteval`](https://newville.github.io/asteval/motivation.html#how-safe-is-asteval) (haven't checked whether it really has all the features)
- [Minimise the security issues of `eval`](https://realpython.com/python-eval-function/#minimizing-the-security-issues-of-eval)

Reply:
How do other SAML clients implement features like this?

`asteval` seems mainly concerned with preventing segfaults; my big concern is more access to the database.

Of the three options, the first one is the one that appeals to me, though we could achieve the same result by writing a DSL for matching SAML attributes -- in fact I hope someone else has written one and we can just copy it. e.g. would [Shibboleth's AttributeFilterPolicyConfiguration](https://wiki.shibboleth.net/confluence/display/IDP4/AttributeFilterPolicyConfiguration) do the job? We could implement a subset of that and gradually improve it as necessary.

Reply:
Attribute Filtering policies are used by IdPs:
> It is about constraining the set of attributes and/or attribute values prior to them being used for some purpose, typically either for passing along to a relying party or less often to limit data passing into the IdP itself from another source. In this way, you can tailor the attributes seen by a specific relying party for a specific subject. 

I think in this case it's better to go with some business rule engines to create a DSL. For example, [this one](https://github.com/venmo/business-rules) looks promising

Reply:
Actually, I found a [question](https://stackoverflow.com/questions/50275634/python-saml-onelogin-equivalent-to-simplesamlphp-filters) on StackOverflow asking about a similar thing. This feature is implemented in [SimpleSAMLphp](https://simplesamlphp.org/docs/stable/simplesamlphp-authproc) and partially in [PySAML2](https://github.com/IdentityPython/pysaml2) but it's not present in [python-saml](https://github.com/onelogin/python-saml) which we're using.

Reply:
Actually, I was wrong, the libraries I mentioned above doesn't allow to filter by attribute values. It seems that only Shibboleth SP allows to it using [AttributeFilter](https://wiki.shibboleth.net/confluence/display/SP3/AttributeFilter).

Reply:
Do you think a library is likely to have an AttributeFilter already configured that they could copy-and-paste into their CM configuration?

Reply:
Right now I'm not even sure that AttributeFilter has the functionality that we need. [The documentation](https://wiki.shibboleth.net/confluence/display/SP3/AttributeFilter) describes the following cases:
- limiting the values of an attribute whose values are required to be from an enumeration (e.g., the eduPersonScopedAffiliation attribute)
- applying automated rules for the acceptance of attribute "scopes" based on SAML metadata
- blocking self asserted personal identification data from known "open" IdPs
- limiting custom attributes intended to be used by only a single IdP